### PR TITLE
DRMAAV2-29: Implementation of Job Session

### PR DESCRIPTION
**Pull request by rampranesh**
#### Issue-ID

* **[DRMAAV2-29](https://pbspro.atlassian.net/browse/DRMAAV2-29)**

#### Problem description
* Implementation of Job Session.

#### Cause / Analysis
* NA

#### Solution description
* NA

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **unit test(s) to my commit**. *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why unit testing is not appropriate** for this case.
- [x] All new and existing automated tests have passed. 
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: src/JobSessionImpl.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+namespace drmaa2 {
+
+const JobList& JobSessionImpl::getJobs(const JobInfo& filter_) {
+	DRMSystem *drms = Singleton<DRMSystem, PBSProSystem>::getInstance();
+	_jobList = drms->getJobs(_jobSessionConnObj, filter_);
+	return _jobList;
+}
+
+const JobArray& JobSessionImpl::getJobArray(const string& jobArrayId_) {
```
Comment (vinodchitrali) [inline src/JobSessionImpl.cpp]: Caller will crash. 

Should not Return local object as reference

File: src/PBSJobArrayImpl.cpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	return _jobId;
+}
+
+JobList& PBSJobArrayImpl::getJobs(void) {
+	PBSConnection pbsconn_(pbs_default(), 0, 0);
+	const Connection &pbsConnPoolObj_ = ConnectionPool::getInstance()->addConnection(pbsconn_);
+	const PBSConnection *pbsCnHolder_ = static_cast<const PBSConnection*>(&pbsConnPoolObj_);
+	struct batch_status *batchResponse_ = NULL;
+	ATTRL* jobAttribs_ = NULL;
+	batchResponse_ = pbs_statjob(pbsCnHolder_->getFd(), (char *)_jobId.c_str(), NULL, (char *)"t");
```
Comment (vinodchitrali) [inline src/PBSJobArrayImpl.cpp]: Should call DRMSystem methods

File: api/cpp-binding/drmaa2.hpp [hunk 3]
```diff
@@ -791,9 +792,9 @@ class ReservationSession {
  */
 class JobSession {
 private:
-	const string contact; /*!< DRMS Contact name */
-	const string sessionName; /*!< Session name */
-	StringList jobCategories; /*!< List of Categories*/
+	const string _contact; /*!< DRMS Contact name */
```
Comment (rampranesh) [inline api/cpp-binding/drmaa2.hpp]: Done!

File: api/cpp-binding/drmaa2.hpp [hunk 4]
```diff
@@ -885,7 +887,7 @@ class JobSession {
 	 *
 	 * @return valid job id returned from the DRMS
 	 */
-	virtual const string& runJob(const JobTemplate& jobTemplate_) = 0;
+	virtual Job& runJob(const JobTemplate& jobTemplate_) const = 0;
```
Comment (rampranesh) [inline api/cpp-binding/drmaa2.hpp]: Done!

File: inc/JobSessionImpl.h [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+#include <string>
+
+#include "drmaa2.hpp"
+#include <ConnectionPool.h>
+
+using namespace std;
+
+namespace drmaa2 {
+
+class JobSessionImpl : public JobSession {
```
Comment (rampranesh) [inline inc/JobSessionImpl.h]: Done!

File: inc/JobSessionImpl.h [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+	/**
+	 * @brief Constructor
+	 *
+	 * @param connObj_ - PBSPro Connection
+	 */
+	JobSessionImpl(const Connection& connObj_) : _jobSessionConnObj(connObj_) {
+
+	}
+public:
+	const Connection &_jobSessionConnObj;
```
Comment (rampranesh) [inline inc/JobSessionImpl.h]: As of now adding a new connection, will change it to getConnection once I have done some modification to ConnectionPool

File: inc/SessionManagerImpl.h [hunk 7]
```diff
@@ -58,6 +59,7 @@ class SessionManagerImpl: public SessionManager {
 	 *
 	 */
 	friend Singleton<SessionManager, SessionManagerImpl> ;
+	map<string, JobSessionImpl> _jobSessionMap;
```
Comment (rampranesh) [inline inc/SessionManagerImpl.h]: Done!

File: src/JobSessionImpl.cpp (hunk 1 above)
Comment (rampranesh) [inline src/JobSessionImpl.cpp]: Done, allocating in heap and returning it!

File: src/PBSJobArrayImpl.cpp (hunk 2 above)
Comment (rampranesh) [inline src/PBSJobArrayImpl.cpp]: Done!

**Comment by rampranesh** [comment]
Hi @vinodchitrali ,
   I have done the needed changes, could you please have a look at the same.

Thanks,
Rampranesh M

File: inc/PBSJobSessionImpl.h [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+public:
+	/**
+	 * @brief Parameterized Constructor
+	 *
+	 * @param contact_ - DRMS contact name
+	 * @param sessionName_ - Session Name
+	 * @param jobCategories_ - Job Categories
+	 *
+	 */
+	PBSJobSessionImpl(const string& contact_,
```
Comment (vinodchitrali) [inline inc/PBSJobSessionImpl.h]: Keep contact as last arg and assign default value. U can make use of default arguments

File: src/PBSProSystem.cpp [hunk 9]
```diff
@@ -235,39 +237,93 @@ Job* PBSProSystem::getJob(const Connection& connection_,
...
+	else if (!jobTemplate_.queueName.empty())
+		destination_.append(jobTemplate_.queueName);
+
+	_attrParse.setAttribute((char *)ATTR_J, (char *)jobIndices_.c_str());
+	ATTRL *attributeList = _attrParse.parseTemplate((void*)&jobTemplate_);
+	jobIdFromDRMS_ = pbs_submit(pbsCnHolder_->getFd(), (struct attropl *) attributeList,
+			(char*) jobTemplate_.remoteCommand.c_str(), (char*) destination_.c_str(), NULL);
+	if(jobIdFromDRMS_) {
+		string jobArrayId_(jobIdFromDRMS_);
+		return new PBSJobArrayImpl(jobArrayId_, jobTemplate_);
```
Comment (vinodchitrali) [inline src/PBSProSystem.cpp]: Some object should take the accounting of this allocated memory and that object should take the responsibility of the releasing this memory. 

Here i see that PBSJobSessionImpl should hold all the objects which should be deleted when PBSJobSessionImpl destructor is called
 

Add doxygen doc about possible memory leak.

File: inc/PBSJobSessionImpl.h [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+#include <string>
+
+#include "drmaa2.hpp"
+#include <ConnectionPool.h>
+
+using namespace std;
+
+namespace drmaa2 {
+
+class PBSJobSessionImpl : public JobSession {
```
Comment (rampranesh) [inline inc/PBSJobSessionImpl.h]: Done!

File: inc/PBSJobSessionImpl.h (hunk 8 above)
Comment (rampranesh) [inline inc/PBSJobSessionImpl.h]: Done!

File: src/PBSProSystem.cpp (hunk 9 above)
Comment (rampranesh) [inline src/PBSProSystem.cpp]: I have added the warning in doxygen header, regarding the deleting of jobs in destructor i found the following in the DRMAA spec so i am leaving it as it. Please let me know your opinion.

> Job and reservation sessions maintain persistent state information (about jobs and reservations created) between application runs.

**Comment by vinodchitrali** [comment]
@PBSPro/pbspro-maintainers , Depended ticket is waiting for PR merge. Please review and merge this code.

File: api/cpp-binding/drmaa2.hpp [hunk 11]
```diff
@@ -885,7 +885,7 @@ class JobSession {
 	 *
 	 * @return valid job id returned from the DRMS
 	 */
-	virtual const string& runJob(const JobTemplate& jobTemplate_) = 0;
+	virtual const Job& runJob(const JobTemplate& jobTemplate_) const = 0;
```
Comment (rampranesh) [inline api/cpp-binding/drmaa2.hpp]: Done!

File: api/cpp-binding/drmaa2.hpp [hunk 12]
```diff
@@ -806,24 +807,23 @@ class JobSession {
...
+	 * @param contact_ - DRMS contact name
 	 *
 	 */
-	JobSession(const string& contact_,
-			const string& sessionName_,
-			const StringList& jobCategories_) :
-					contact(contact_),
-					sessionName(sessionName_),
-					jobCategories(jobCategories_) {
+	JobSession( const string& sessionName_,
```
Comment (rampranesh) [inline api/cpp-binding/drmaa2.hpp]: contact_ can be passed as a default argument if the user does not specify contact name. That is the reason we have changes the arguments order.

**Comment by rampranesh** [comment]
@arungrover  I have addressed your comments please have a look at the changes.

**Comment by arungrover** [comment]
please rebase and squash

**Comment by rampranesh** [comment]
Hi @arun,
      I have rebased and squashed my commits.

Thanks,
Rampranesh M

**Pull request merged**